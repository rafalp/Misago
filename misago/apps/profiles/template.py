from datetime import timedelta
from django.conf import settings
from django.template import RequestContext as DjangoRequestContext
from django.utils import timezone
from django.utils.importlib import import_module
from misago.models import User

def RequestContext(request, context=None):
    if not context:
        context = {}
    context['fallback'] = request.path
        
    # Find out if we ignore or follow this user
    context['follows'] = False
    context['ignores'] = False
    if request.user.is_authenticated() and request.user.pk != context['profile'].pk:
        context['follows'] = request.user.is_following(context['profile'])
        context['ignores'] = request.user.is_ignoring(context['profile'])
    
    # Find out if this user allows us to see his activity
    if request.user.pk != context['profile'].pk:
        if context['profile'].hide_activity == 2:
            context['hidden'] = True
        if context['profile'].hide_activity == 1:
            context['hidden'] = context['profile'].is_following(request.user)
    else:
        context['hidden'] = False

    # Find out if this user is online:
    if request.user.pk != context['profile'].pk:
        try:
            context['online'] = context['profile'].sessions.filter(admin=False).filter(last__gt=(timezone.now() - timedelta(minutes=10))).order_by('-last')[0:1][0]
        except IndexError:
            context['online'] = False
    else:
        # Fake "right now" time
        context['online'] = {'last': timezone.now()}

    # Sync member
    if context['profile'].sync_profile():
        context['profile'].save(force_update=True)

    context['tabs'] = []
    for extension in settings.PROFILE_EXTENSIONS:
        profile_module = import_module(extension + '.profile')
        try:
            append_links = profile_module.register_profile_extension(request)
            if append_links:
                for link in append_links:
                    link = list(link)
                    token = link[0][link[0].find('_') + 1:]
                    context['tabs'].append({
                                            'route': link[0],
                                            'active': context['tab'] == token,
                                            'name': link[1],
                                            })
        except AttributeError:
            pass
    return DjangoRequestContext(request, context)
