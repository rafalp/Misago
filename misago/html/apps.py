from django.apps import AppConfig


class MisagoHTMLConfig(AppConfig):
    name = "misago.html"
    label = "misago_HTML"
    verbose_name = "Misago HTML"
