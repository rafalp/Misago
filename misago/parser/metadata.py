from typing import Iterable

from django.conf import settings
from django.contrib.auth import get_user_model

from ..core.utils import slugify
from .context import ParserContext
from .hooks import (
    get_ast_metadata_users_queryset_hook,
    update_ast_metadata_hook,
    update_ast_metadata_from_node_hook,
    update_ast_metadata_users_hook,
)

User = get_user_model()


def create_ast_metadata(
    context: ParserContext,
    ast: list[dict],
) -> dict:
    metadata = {
        "outbound-links": set(),
        "usernames": set(),
        "users": {},
        "posts": {
            "ids": set(),
            "objs": {},
        },
    }

    return update_ast_metadata_hook(_update_ast_metadata_action, context, ast, metadata)


def _update_ast_metadata_action(
    context: ParserContext,
    ast: list[dict],
    metadata: dict,
) -> dict:
    for ast_node in ast:
        update_ast_metadata_from_node(context, ast_node, metadata)

    update_ast_metadata_users(context, metadata)

    return metadata


def update_ast_metadata_from_node(
    context: ParserContext,
    ast_node: dict,
    metadata: dict,
) -> None:
    update_ast_metadata_from_node_hook(
        _update_ast_metadata_from_node_action, context, ast_node, metadata
    )


def _update_ast_metadata_from_node_action(
    context: ParserContext,
    ast_node: dict,
    metadata: dict,
) -> None:
    if ast_node["type"] == "mention":
        metadata["usernames"].add(slugify(ast_node["username"]))

    elif ast_node["type"] == "quote-bbcode":
        if ast_node["author"]:
            metadata["usernames"].add(slugify(ast_node["author"]))
        if ast_node["post"]:
            metadata["posts"]["ids"].add(ast_node["post"])

    elif ast_node["type"] in ("auto-link", "auto-url", "url", "url-bbcode"):
        if not context.forum_address.is_inbound_link(ast_node["href"]):
            metadata["outbound-links"].add(ast_node["href"])

    if ast_node.get("children"):
        for child_node in ast_node["children"]:
            update_ast_metadata_from_node(context, child_node, metadata)

    if ast_node.get("items"):
        for child_node in ast_node["items"]:
            update_ast_metadata_from_node(context, child_node, metadata)

    if ast_node.get("lists"):
        for child_node in ast_node["lists"]:
            update_ast_metadata_from_node(context, child_node, metadata)


def update_ast_metadata_users(context: ParserContext, metadata: dict) -> None:
    update_ast_metadata_users_hook(_update_ast_metadata_users_action, context, metadata)


def _update_ast_metadata_users_action(context: ParserContext, metadata: dict) -> None:
    if not metadata["usernames"]:
        return

    usernames = sorted(metadata["usernames"])
    if len(usernames) > settings.MISAGO_PARSER_MAX_USERS:
        return

    if usernames:
        queryset = get_ast_metadata_users_queryset(context, usernames)
        for user in queryset:
            metadata["users"][user.slug] = user


def get_ast_metadata_users_queryset(
    context: ParserContext, usernames: list[str]
) -> Iterable[User]:
    return get_ast_metadata_users_queryset_hook(
        _get_ast_metadata_users_queryset_action, context, usernames
    )


def _get_ast_metadata_users_queryset_action(
    context: ParserContext, usernames: list[str]
) -> Iterable[User]:
    return User.objects.filter(slug__in=usernames)
