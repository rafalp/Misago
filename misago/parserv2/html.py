from markdown_it import MarkdownIt
from markdown_it.token import Token


def render_tokens_to_html(parser: MarkdownIt, tokens: list[Token]) -> str:
    return parser.renderer.render(tokens, parser.options, {}).strip()
