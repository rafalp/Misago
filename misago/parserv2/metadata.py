from markdown_it.token import Token

from .hooks import get_tokens_metadata_hook


def get_tokens_metadata(tokens: list[Token]) -> dict:
    return get_tokens_metadata_hook(_get_tokens_metadata_action, tokens)


def _get_tokens_metadata_action(tokens: list[Token]) -> dict:
    metadata = {}

    if attachments := get_attachments_metadata(tokens):
        metadata["attachments"] = attachments

    if get_highlight_code_metadata(tokens):
        metadata["highlight_code"] = True

    if posts := get_quoted_posts_metadata(tokens):
        metadata["posts"] = posts

    if mentions := get_mentions_metadata(tokens):
        metadata["mentions"] = mentions

    return metadata


def get_attachments_metadata(tokens: list[Token]) -> list[int]:
    attachments: set[int] = set()
    get_metadata_recursive(tokens, "attachment", "attachment", attachments)
    return sorted(attachments)


def get_highlight_code_metadata(tokens: list[Token]) -> list[str]:
    syntax: set[str] = set()
    get_metadata_recursive(tokens, "fence", "syntax", syntax)
    get_metadata_recursive(tokens, "code_bbcode", "syntax", syntax)
    return sorted(syntax)


def get_quoted_posts_metadata(tokens: list[Token]) -> list[int]:
    posts: set[int] = set()
    get_metadata_recursive(tokens, "quote_bbcode_open", "post", posts)
    return sorted(posts)


def get_mentions_metadata(tokens: list[Token]) -> list[str]:
    mentions: set[str] = set()
    get_metadata_recursive(tokens, "mention", "slug", mentions)
    return sorted(mentions)


def get_metadata_recursive(
    tokens: list[Token], token_type: str, meta: str, data: set
) -> set:
    for token in tokens:
        if token.type == token_type and token.meta and token.meta.get(meta):
            data.add(token.meta[meta])

        if token.children:
            get_metadata_recursive(token.children, token_type, meta, data)
