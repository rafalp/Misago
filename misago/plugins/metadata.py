import re
from dataclasses import dataclass
from importlib import import_module
from pathlib import Path
from types import ModuleType
from typing import Dict, List, Optional
from urllib.parse import urlparse

from django.conf import settings
from django.utils.translation import pgettext_lazy

from .manifest import MisagoPlugin


@dataclass(frozen=True)
class PluginMetadataUrl:
    url: str
    netloc: str


@dataclass(frozen=True)
class PluginMetadata:
    package: str
    dirname: str
    has_manifest: bool
    manifest_error: Optional[str]

    name: Optional[str]
    description: Optional[str]
    license: Optional[str]
    icon: Optional[str]
    color: Optional[str]
    version: Optional[str]
    author: Optional[str]
    homepage: Optional[PluginMetadataUrl]
    sponsor: Optional[PluginMetadataUrl]
    help: Optional[PluginMetadataUrl]
    bugs: Optional[PluginMetadataUrl]
    repo: Optional[PluginMetadataUrl]

    @property
    def has_urls(self) -> bool:
        return bool(
            self.homepage or self.sponsor or self.help or self.bugs or self.repo
        )


class PluginsMetadataLoader:
    initialized: bool
    plugins: List[str]
    metadata: Dict[str, PluginMetadata]

    def __init__(self, plugins: List[str]):
        self.initialized = False
        self.plugins = plugins
        self.metadata = {}

    def get_metadata(self) -> Dict[str, PluginMetadata]:
        if not self.initialized:
            self.metadata = self.build_metadata(self.plugins)
            self.initialized = True

        return self.metadata

    def build_metadata(self, plugins: List[str]) -> Dict[str, PluginMetadata]:
        metadata: Dict[str, PluginMetadata] = {}
        for plugin_package in plugins:
            metadata[plugin_package] = self.build_plugin_metadata(plugin_package)
        return metadata

    def build_plugin_metadata(self, plugin_package: str) -> PluginMetadata:
        package_obj = import_module(plugin_package)

        manifest: Optional[MisagoPlugin] = None
        manifest_error: Optional[str] = None

        try:
            manifest_module = import_module(f"{plugin_package}.misago_plugin")
            manifest_attr = getattr(manifest_module, "manifest", None)
            if manifest_attr is not None and not isinstance(
                manifest_attr, MisagoPlugin
            ):
                raise TypeError(
                    pgettext_lazy(
                        "plugins metadata",
                        "Plugin manifest is of an unexpected type '%(manifest_type)s' (expected MisagoPlugin instance)",
                    )
                    % {"manifest_type": type(manifest_attr).__name__}
                )
            manifest = manifest_attr
        except Exception as exc:
            manifest_error = f"{type(exc).__name__}: {exc}"

        return create_plugin_metadata(
            plugin_package,
            package_obj,
            manifest,
            manifest_error,
        )


def create_plugin_metadata(
    plugin_package: str,
    package_obj: ModuleType,
    manifest: Optional[MisagoPlugin],
    manifest_error: Optional[str],
) -> PluginMetadata:
    has_manifest = False

    name: Optional[str] = None
    description: Optional[str] = None
    license: Optional[str] = None
    icon: Optional[str] = None
    color: Optional[str] = None
    version: Optional[str] = None
    author: Optional[str] = None
    homepage: Optional[PluginMetadataUrl] = None
    sponsor: Optional[PluginMetadataUrl] = None
    help: Optional[PluginMetadataUrl] = None
    bugs: Optional[PluginMetadataUrl] = None
    repo: Optional[PluginMetadataUrl] = None

    if manifest:
        has_manifest = True
        name = clean_plugin_str(manifest.name, 100)
        description = clean_plugin_str(manifest.description, 250)
        license = clean_plugin_str(manifest.license, 50)
        icon = clean_plugin_icon(manifest.icon)
        color = clean_plugin_color(manifest.color)
        version = clean_plugin_str(manifest.version, 50)
        author = clean_plugin_str(manifest.author, 150)
        homepage = clean_plugin_url(manifest.homepage)
        sponsor = clean_plugin_url(manifest.sponsor)
        help = clean_plugin_url(manifest.help)
        bugs = clean_plugin_url(manifest.bugs)
        repo = clean_plugin_url(manifest.repo)

    return PluginMetadata(
        package=plugin_package,
        dirname=Path(package_obj.__file__).parent.parent.name,
        has_manifest=has_manifest,
        manifest_error=manifest_error,
        name=name,
        description=description,
        license=license,
        icon=icon,
        color=color,
        version=version,
        author=author,
        homepage=homepage,
        sponsor=sponsor,
        help=help,
        bugs=bugs,
        repo=repo,
    )


def clean_plugin_str(value: Optional[str], max_length: int) -> Optional[str]:
    if isinstance(value, str):
        return (value.strip()[:max_length].strip()) or None
    return None


URL_RE = re.compile("^https?://[A-Za-z0-9-_/?=&#.]+$")


def clean_plugin_url(value: Optional[str]) -> Optional[PluginMetadataUrl]:
    if isinstance(value, str) and URL_RE.match(value):
        try:
            url = urlparse(value)
            if not url:
                return None

            return PluginMetadataUrl(
                url=value,
                netloc=url.netloc,
            )
        except ValueError:
            return None
    return None


ICON_RE = re.compile("^fa(r|s)? fa-[a-z-]+$")


def clean_plugin_icon(value: Optional[str]) -> Optional[str]:
    if isinstance(value, str) and ICON_RE.match(value):
        return value
    return None


COLOR_RE = re.compile("^#[0-9a-f]{6}$", re.IGNORECASE)


def clean_plugin_color(value: Optional[str]) -> Optional[str]:
    if isinstance(value, str) and COLOR_RE.match(value):
        return value.upper()
    return None


plugins_metadata = PluginsMetadataLoader(settings.INSTALLED_PLUGINS)
