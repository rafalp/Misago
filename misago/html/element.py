from html import escape


def html_element(
    tag: str,
    children: str | None = None,
    attrs: dict | None = None,
) -> str:
    html = tag

    if attrs:
        html += _html_attributes(attrs)

    if children is None:
        return f"<{html} />"

    return f"<{html}>{children}</{tag}>"


def _html_attributes(attrs: dict) -> str:
    html = ""
    for attr, value in attrs.items():
        if value is True:
            html += " " + attr
        elif value is not None:
            html += f' {attr}="{escape(value)}"'
    return html
