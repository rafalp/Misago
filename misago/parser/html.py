from html import escape

from django.utils.translation import pgettext

from ..core.utils import slugify
from .context import ParserContext
from .exceptions import AstError
from .hooks import complete_markup_html_hook, render_ast_node_to_html_hook
from .urls import clean_href

SPOILER_SUMMARY = "<spoiler-summary-message>"


def render_ast_to_html(context: ParserContext, ast: list[dict], metadata: dict) -> str:
    html = ""
    for ast_node in ast:
        node_html = render_ast_node_to_html(context, ast_node, metadata)
        if node_html:
            html += node_html
    return html


def render_ast_node_to_html(
    context: ParserContext, ast_node: dict, metadata: dict
) -> str:
    return render_ast_node_to_html_hook(
        _render_ast_node_to_html_action, context, ast_node, metadata
    )


def _render_ast_node_to_html_action(
    context: ParserContext, ast_node: dict, metadata: dict
) -> str:
    ast_type = ast_node["type"]

    if ast_type in ("heading", "heading-setex"):
        html_tag = f"h{ast_node['level']}"
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<{html_tag}>{children}</{html_tag}>"

    if ast_type == "list":
        html_tag = "ol" if ast_node["ordered"] else "ul"
        children = render_ast_to_html(context, ast_node["items"], metadata)
        return f"<{html_tag}>{children}</{html_tag}>"

    if ast_type == "list-item":
        text = render_ast_to_html(context, ast_node["children"], metadata)
        children = render_ast_to_html(context, ast_node["lists"], metadata)
        return f"<li>{text}{children}</li>"

    if ast_type in ("code", "code-bbcode"):
        if ast_node["syntax"]:
            html_class = f" class=\"language-{ast_node['syntax']}\""
        else:
            html_class = ""
        return f"<pre{html_class}><code>{escape(ast_node['code'])}</code></pre>"

    if ast_type == "code-indented":
        return f"<pre><code>{escape(ast_node['code'])}</code></pre>"

    if ast_type == "code-inline":
        return f"<code>{escape(ast_node['code'])}</code>"

    if ast_type == "quote":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<blockquote>{children}</blockquote>"

    if ast_type == "quote-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)

        if not ast_node["author"]:
            return f"<blockquote>{children}</blockquote>"

        heading = escape(ast_node["author"])

        return (
            '<aside class="quote-block">'
            f'<div class="quote-heading" data-noquote="1">{heading}</div>'
            f'<blockquote class="quote-body">{children}</blockquote>'
            "</aside>"
        )

    if ast_type == "spoiler-bbcode":
        if ast_node["summary"]:
            summary = escape(ast_node["summary"])
        else:
            summary = SPOILER_SUMMARY
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<details><summary>{summary}</summary>{children}</details>"

    if ast_type == "paragraph":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<p>{children}</p>"

    if ast_type in ("emphasis", "emphasis-underscore"):
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<em>{children}</em>"

    if ast_type in ("strong", "strong-underscore"):
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<strong>{children}</strong>"

    if ast_type in ("strikethrough", "strikethrough-bbcode"):
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<del>{children}</del>"

    if ast_type == "bold-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<b>{children}</b>"

    if ast_type == "italics-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<i>{children}</i>"

    if ast_type == "underline-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<u>{children}</u>"

    if ast_type == "superscript-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<sup>{children}</sup>"

    if ast_type == "subscript-bbcode":
        children = render_ast_to_html(context, ast_node["children"], metadata)
        return f"<sub>{children}</sub>"

    if ast_type in ("thematic-break", "thematic-break-bbcode"):
        return "<hr />"

    if ast_type in ("image", "image-bbcode"):
        src = escape(clean_href(ast_node["src"]))
        alt = escape(ast_node["alt"]) if ast_node["alt"] else ""
        return f'<img src="{src}" alt="{alt}" />'

    if ast_type in ("url", "url-bbcode"):
        children = render_ast_to_html(context, ast_node["children"], metadata)
        href = escape(clean_href(ast_node["href"]))
        rel = "external nofollow noopener"
        return f'<a href="{href}" rel="{rel}" target="_blank">{children or href}</a>'

    if ast_type in ("auto-link", "auto-url"):
        href = escape(clean_href(ast_node["href"]))
        rel = "external nofollow noopener"
        if ast_node.get("image"):
            return f'<img src="{href}" alt="" />'
        else:
            return f'<a href="{href}" rel="{rel}" target="_blank">{href}</a>'

    if ast_type == "mention":
        username = slugify(ast_node["username"])
        if username not in metadata["users"]:
            return escape("@" + ast_node["username"])

        user = metadata["users"][username]
        return f'<a href="{user.get_absolute_url()}">@{escape(user.username)}</a>'

    if ast_type == "escape":
        return escape(ast_node["character"])

    if ast_type == "line-break":
        return "<br />"

    if ast_type == "text":
        return escape(ast_node["text"])

    raise AstError(f"Unknown AST node type: {ast_type}")


def complete_markup_html(html: str, **kwargs) -> str:
    return complete_markup_html_hook(_complete_markup_html_action, html, **kwargs)


def _complete_markup_html_action(html: str, **kwargs) -> str:
    html = complete_markup_html_spoiler_summary(html)
    return html


def complete_markup_html_spoiler_summary(html: str) -> str:
    if SPOILER_SUMMARY in html:
        html = html.replace(
            SPOILER_SUMMARY, pgettext("spoiler summary", "Reveal spoiler")
        )

    return html
