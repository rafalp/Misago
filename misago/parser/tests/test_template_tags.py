from django.template import Context, Template

from ..html import complete_markup_html, render_ast_to_html
from ..metadata import create_ast_metadata


def render(template_str, context: dict | None = None):
    base_template = "{%% load misago_markup %%} %s"
    context = Context(context or {})
    template = Template(base_template % template_str)
    return template.render(context).strip()


def test_complete_markup_template_tag_replaces_default_spoiler_summary(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("[spoiler]Hello world![/spoiler]")
    metadata = create_ast_metadata(parser_context, ast)
    html = render_ast_to_html(parser_context, ast, metadata)
    assert snapshot == render("{% completemarkup html %}", {"html": html})
