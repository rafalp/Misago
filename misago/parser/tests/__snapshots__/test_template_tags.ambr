# serializer version: 1
# name: test_complete_markup_template_tag_replaces_default_spoiler_summary
  '<details><summary>Reveal spoiler</summary><p>Hello world!</p></details>'
# ---
